src/device.jl

```diff
@@ -18,9 +25,11 @@ function device_reset(dev::Integer)
     for p in todelete
         delete!(cuda_ptrs, p)
     end
-    # Reset the device
-    device(dev)
-    rt.cudaDeviceReset()
+
+    # no need to reset the entire device, just reset the primary context
+    pctx = CuPrimaryContext(dev)
+    CUDAdrv.reset(pctx)
```

Comment:
There is still a problem with ordering here. I am seeing a seqfault right now where the following happens:

```
TRACE: cuDevicePrimaryCtxReset(pctx.dev=0) = SUCCESS(0)
TRACE: cuDevicePrimaryCtxRetain(handle=Base.RefValue{Ptr{Void}}(Ptr{Void} @0x00007f3f4a860550), pctx.dev=0) = SUCCESS(0)
TRACE: Finalizing CuModule at Ptr{Void} @0x00007f3f50d012f0
TRACE: cuModuleUnload(mod=CUDAdrv.CuModule(Ptr{Void} @0x00000000043435a0,CUDAdrv.CuContext(Ptr{Void} @0x0000000003b2dd20,false))) =
signal (11): Segmentation fault
```

Which means we will need to invalidate all objects that hold a reference to the context. Also of note is that the after a reset -> retain the pointer is still the same.